device-connectors/src/testflinger_device_connectors/devices/zapper_kvm/__init__.py

```diff
@@ -50,18 +60,116 @@ def _validate_configuration(
+            username = self.job_data.get("test_data", {}).get(
                 "test_username", "ubuntu"
-            ),
-            "password": self.job_data.get("test_data", {}).get(
+            )
+            password = self.job_data.get("test_data", {}).get(
                 "test_password", "ubuntu"
-            ),
+            )
+            retries = self.job_data["provision_data"].get("robot_retries", 1)
+
+        provisioning_data = {
+            "url": url,
+            "username": username,
+            "password": password,
             "autoinstall_conf": self._get_autoinstall_conf(),
             "reboot_script": self.config["reboot_script"],
             "device_ip": self.config["device_ip"],
             "robot_tasks": self.job_data["provision_data"]["robot_tasks"],
+            "robot_retries": retries,
+            "cmdline_append": self.job_data["provision_data"].get(
+                "cmdline_append", ""
+            ),
+            "skip_download": self.job_data["provision_data"].get(
+                "skip_download", False
+            ),
         }
 
         return ((), provisioning_data)
+
+    def _post_run_actions(self, args):
+        super()._post_run_actions(args)
+
+        if "alloem_url" in self.job_data["provision_data"]:
+            self._change_password("ubuntu", "u")
+            self._copy_ssh_id()
+            self._run_oem_script(args)
+
+    def _run_oem_script(self, args):
+        """
+        If "alloem_url" was in scope, the Zapper only restored
+        the OEM reset partition. The usual oemscript will take care
+        of the rest.
+        """
+
+        if not self.job_data["provision_data"].get("url"):
+            logger.warning(
+                "Provisioned with base `alloem` image, no test URL specified."
+            )
+            return
+
+        oem = self.job_data["provision_data"].get("oem")
+        oemscript = {
+            "hp": HPOemScript,
+            "dell": DellOemScript,
+            "lenovo": LenovoOemScript,
+        }.get(oem, OemScript)(args.config, args.job_data)
+
+        oemscript.provision()
+
+    def _copy_ssh_id(self):
+        """Copy the ssh id to the device"""
+        try:
+            test_username = self.job_data.get("test_data", {}).get(
+                "test_username", "ubuntu"
+            )
+            test_password = self.job_data.get("test_data", {}).get(
+                "test_password", "ubuntu"
+            )
+        except AttributeError:
+            test_username = "ubuntu"
+            test_password = "ubuntu"
+
+        cmd = [
+            "sshpass",
+            "-p",
+            test_password,
+            "ssh-copy-id",
+            "-o",
+            "StrictHostKeyChecking=no",
+            "-o",
+            "UserKnownHostsFile=/dev/null",
+            f"{test_username}@{self.config['device_ip']}",
+        ]
+        subprocess.check_output(cmd, stderr=subprocess.STDOUT, timeout=60)
```

Comment:
Have you looked at these calls to check_output to ensure you are ok with the exceptions that could get raised? I'm finding it much nicer to call subprocess.run in a lot of cases (if possible), which still handling things like timeout. And either raising a more predictable exception, or even returning the return_code and output is often more useful than using check_output()

Reply:
In this case I tend to prefer `check_output` since I care about the result only when it fails. I wasn't handling at all the exceptions though, thanks for noticing. Updated with proper logging and exception.